application/client/src/app/ui/tabs/sources/multiplefiles/structure/component.ts

```diff
@@ -0,0 +1,191 @@
+import {
+    Component,
+    ViewChild,
+    ChangeDetectorRef,
+    AfterViewInit,
+    AfterContentInit,
+    Input,
+    ViewEncapsulation,
+    Output,
+    EventEmitter,
+} from '@angular/core';
+import { Ilc, IlcInterface } from '@env/decorators/component';
+import { ChangesDetector } from '@ui/env/extentions/changes';
+import { MatSort } from '@angular/material/sort';
+import { MatTableDataSource } from '@angular/material/table';
+import { FileHolder } from '../file.holder';
+import { MatTable } from '@angular/material/table';
+import { bytesToStr } from '@env/str';
+
+export const COLUMNS = {
+    type: 'type',
+    name: 'name',
+    path: 'path',
+    size: 'size',
+    modification_date: 'modification date',
+};
+
+@Component({
+    selector: 'app-tabs-source-multiplefiles-structure',
+    templateUrl: './template.html',
+    styleUrls: ['./styles.less'],
+    encapsulation: ViewEncapsulation.None,
+})
+@Ilc()
+export class TabSourceMultipleFilesStructure
+    extends ChangesDetector
+    implements AfterContentInit, AfterViewInit
+{
+    columns: string[] = [
+        COLUMNS.type,
+        COLUMNS.name,
+        COLUMNS.path,
+        COLUMNS.size,
+        COLUMNS.modification_date,
+    ];
+
+    @Input() files!: FileHolder[];
+    @Output() context: EventEmitter<null | FileHolder[]> = new EventEmitter();
+
+    @ViewChild(MatSort) sort!: MatSort;
+    @ViewChild(MatTable) table!: MatTable<any>;
+
+    public data!: MatTableDataSource<FileHolder>;
+
+    constructor(cdRef: ChangeDetectorRef) {
+        super(cdRef);
+    }
+
+    public ngAfterContentInit() {
+        this.data = new MatTableDataSource<FileHolder>(this.files);
+    }
+
+    public ngAfterViewInit() {
+        this.data.sort = this.sort;
+    }
+
+    public ngModificationDate(time: number): string {
+        return new Date(time).toLocaleString('en-GB');
+    }
+
+    public ngOnSortChange() {
+        this.detectChanges();
+        this.table.renderRows();
+    }
+
+    public ngOnRowSelect(file: FileHolder) {
+        file.reverseSelect();
+        this.context.next(null);
+    }
+
+    public ngSize(size: number): string {
+        return bytesToStr(size);
+    }
+
+    public ngContextMenu(event: any, file?: FileHolder) {
+        event.stopImmediatePropagation();
```

Comment:
is it really needed? as Soon as it's listener of `contextmenu` event, I think no need to stop it.

Reply:
We use `ngContextMenu` in both `<table>` (line 1) and `<tr>` (line 40). If we were to call the context menu on the `<tr>`, it would be called on the table row first and then on the table as well, because the table itself is the parent.